### ISSUE TITLE
getPolyHeight(long ref, float[] pos)

### ISSUE THREAD

**Issue by mitm001** [ppiastucki/recast4j#18]
I found that when using a large crowd that moves along the edge of the navigation mesh, that a spatial can be pushed off the mesh when it enters a triangle that is **very** small or where there are two triangles that share a single vertex and the angle between them forms a "V" where the space inside the "V" is off mesh.

It looks like this happens because the velocity of the spatial along with the distance left inside the navigation mesh triangle to travel is to small for a course correction to occur because the spatials following it are blocking its path. This eventually leads to it being left no place to go except off mesh which causes it to throw the exception in getPolyHeight(long ref, float[] pos). 

This may never happen in game because unless you are moving an army of spatials there would not be enough spatials following to cause a problem.

Is it possible that when this exception happens, rather than throwing an exception, try to warp the spatial back onto the mesh at the nearest valid position and replan the path? 

After the crowd passes it would have an open path.

I hope the explanation is clear.

**Comment by ppiastucki** [comment]
If you spot a problem in the underlying algorithms that potentially affects the original library in C I really encourage you to post your findings here: https://groups.google.com/forum/#!forum/recastnavigation.  You can usually get help and learn about the known limitations or bugs in recast and detour quickly there. I try to keep the library as close to the original implementation as possible, so even if it were a bug I would rather wait for the upstream fix than try to come up with a recast4j-specific solution. 
On the other hand, if this issue can only be reproduced with recast4j then definitely something was ported incorrectly and needs to be fixed here :)
Would you be able to test your scenario using detour-crowd in C ?

**Comment by mitm001** [comment]
There is a JNI port of recast made for the engine I use but it was from 2014 and did not include anything other than mesh generation.

I don't know C++ or JNI enough to pull it off.

I can work around the problem by tweaking the mesh but hadn't thought about this being something in recast itself which is probably 100% correct.

Ill dig around and see what I find.

I can give you some parameters on how I implement this to let you duplicate the problem if you want.

**Comment by mitm001** [comment]
Ok, I found this problem is caused by the Y position and wrote a fix for it.

https://groups.google.com/forum/#!msg/recastnavigation/idkwKdgChlA/l-7tT22UydYJ;context-place=forum/recastnavigation

When I examined [movePosition](https://github.com/recastnavigation/recastnavigation/blob/master/DetourCrowd/Source/DetourPathCorridor.cpp#L439) in DetourPathCorridor.cpp it looks like to me he never expects dtStatusSucceed(status)) to return false or doesn't care if it does and if [getPolyHeight() ](https://github.com/recastnavigation/recastnavigation/blob/7cca61dc417ea2e7dacad07f313eede08ad711d2/Detour/Source/DetourNavMeshQuery.cpp#L654) returns DT_FAILURE | DT_INVALID_PARAM I don't see how that gets handled.  

In you implementation you choose to throw a exception. I don't know why its expected to shutdown the game for this so this is my work around.
[NavMeshQuery.java](https://github.com/ppiastucki/recast4j/blob/59112f0d24600f11349191fad7227c0f1850a6b5/detour/src/main/java/org/recast4j/detour/NavMeshQuery.java#L532)
I added a new m_posY[] array and in the getPolyHeight(long ref, float[] pos) method changed from a exception to warping the Y point back into the mesh.
```
 //		throw new IllegalArgumentException("Invalid ref " + ref + " pos " + Arrays.toString(pos));
                ClosesPointOnPolyResult closestPointOnPoly = closestPointOnPoly(ref, pos);
                vCopy(m_posY, closestPointOnPoly.getClosest());
                System.out.println(closestPointOnPoly.isPosOverPoly());
                return m_posY[1];

```

It works and so far I am not seeing anything adverse from it. If you see something that may lead to a problem or there is a cleaner or preferred way to handle this please let me know.

**Comment by ppiastucki** [comment]
You do not need to shut down the application on the exception, simply catch it and perform some fallback action. 
The original implementation (https://github.com/recastnavigation/recastnavigation/blob/master/Detour/Source/DetourNavMeshQuery.cpp#L654) currently does not return any value in case of DT_INVALID_PARAM. If your fix gets merged into the C++ lib I will happily merge it to recast4j as well, otherwise I prefer to remain consistent with what is in C++ code.
I am wondering why the status is not checked in DetourPathCorridor.cpp while it is checked in a few places in DetourNavMeshQuery.cpp. It might be a bug or an intentional omission. I guess we can change the movePosition method to catch and disregard the exception as long as there is no check in C++.

**Comment by mitm001** [comment]
> I prefer to remain consistent with what is in C++ code.

I agree completely with that approach and was only showing what I did to fix it. 

> why the status is not checked in DetourPathCorridor.cpp 

I think the reason is no action needs to be taken. 

I tested by catching the exception and applying my fix vs catching the exception and doing nothing and there is no adverse action either way. Calling closestPointOnPoly is just a waste of processing power IMO. In more than half to 3/4 the time it doesn't even change the results. When it does return something different, 99% of the time its not even above the mesh and the difference in position is so miniscule that's its not worth changing. 

Heres a sample of the output to show how small a difference it made,
```
Invalid ref [-147.32025, 35.05475, -562.08984]
Above Poly false
Corrected ref [-147.32025, 35.05473, -562.08984] pos changed for: org.recast4j.detour.crowd.PathCorridor@79aaf718 5/5
```
This object had 5 failures in the end and returned a different position 5 times but look a Y pos and its to small to worry about.

> I  guess we can change the movePosition method to catch and disregard the exception as long as there is no check in C++.

I think this is the best soloution,
